### PR TITLE
Common function for library address validation

### LINKED ISSUE #10299: Inconsistent library address validations between CLI and Standard JSON
Checks applied to the value of `--library` option in CLI are not exactly the same as the ones for `libraries` in Standard JSON. For example only CLI checks address checksum. The validations should be reviewed and unified.

### PR DESCRIPTION
**Pull request by davidBar-On**
Fixes #10299

Suggested fix for the issue - added `validateAddress()` function and related tests.  The function was added to **new** `Address.h/cpp` as I did find an existing header file where to add the new function.

The new function includes the checksum check, which is new for JSON input.  Hopefully this will not cause compile errors for existing files. 

(Sorry for the two redundant commits, as I was not aware about `check_style.sh` and its required use.)

### CODE REVIEW

File: libsolutil/Address.cpp
Comment (matheusaaguiar) [inline libsolutil/Address.cpp]: We are allowed to use namespace `std` in `cpp` files. You can add it to the top of the file and remove the qualifier.

```suggestion
Result<h160> validateAddress(string _addrString)
{
	string message;
```

File: libsolutil/Address.cpp
Comment (matheusaaguiar) [inline libsolutil/Address.cpp]:
```suggestion
			message = 
				"Invalid library address length " +
				std::to_string(addrString.length()) + 
				" instead of 40 characters.";
```

File: libsolutil/Address.cpp
Comment (matheusaaguiar) [inline libsolutil/Address.cpp]:
```suggestion

```

**Review by matheusaaguiar — COMMENTED** [review]
Hi @davidBar-On ! Thanks for the PR.
I think it is already good, but there are some style adjustments I left as suggestions.
> The function was added to new Address.h/cpp as I did find an existing header file where to add the new function.

I think that `libsolutil/CommonData.h/cpp` would be a good place, since we also have `passesAddressChecksum` there, for example.

Don't worry about the fix commits, we can squash them later before merging.

File: libsolutil/Address.cpp
Comment (davidBar-On) [reply libsolutil/Address.cpp]: Changed.

File: libsolutil/Address.cpp
Comment (davidBar-On) [reply libsolutil/Address.cpp]: Changed.

File: libsolutil/Address.cpp
Comment (davidBar-On) [reply libsolutil/Address.cpp]: Changed.

**Comment by davidBar-On** [comment]
>>The function was added to new Address.h/cpp as I did find an existing header file where to add the new function.

>I think that `libsolutil/CommonData.h/cpp` would be a good place, since we also have `passesAddressChecksum` there, for example.

There is a problem with this approach that **I am not sure how to fix**. `h160` is defined in `FixedHash.h`.  However, including `FixedHash.h` in `CommonData.h` causes issues as the one below.  I assume this is because `FixedHash.h` is including `CommonDat.h`, so there is some kind of cyclic definitions.

```
[  5%] Building CXX object libsolutil/CMakeFiles/solutil.dir/CommonData.cpp.o
In file included from /home/dbo/Tools/solidity/libsolutil/FixedHash.h:28,
                 from /home/dbo/Tools/solidity/libsolutil/CommonData.h:30,
                 from /home/dbo/Tools/solidity/libsolutil/CommonData.cpp:24:
/home/dbo/Tools/solidity/libsolutil/Numeric.h: In function ‘std::string solidity::toHex(solidity::u256)’:
/home/dbo/Tools/solidity/libsolutil/Numeric.h:138:15: error: ‘toHex’ is not a member of ‘solidity::util’; did you mean ‘solidity::toHex’?
  138 |  return util::toHex(toBigEndian(val));
      |               ^~~~~
/home/dbo/Tools/solidity/libsolutil/Numeric.h:136:20: note: ‘solidity::toHex’ declared here
  136 | inline std::string toHex(u256 val)
```

**Comment by matheusaaguiar** [comment]
@davidBar-On , I hadn't noticed that cyclic dependency. In that case, I think that your approach, adding `Address.h/cpp`, might be the best here. Can you squash the commits and rebase on develop, please?

**Comment by davidBar-On** [comment]
Done - squashed and rebased.

File: libsolutil/Address.cpp
Comment (matheusaaguiar) [inline libsolutil/Address.cpp]:
```suggestion
	else
	{
```

File: libsolutil/Address.cpp
Comment (matheusaaguiar) [inline libsolutil/Address.cpp]:
```suggestion
		else
		{
```

**Review by matheusaaguiar — DISMISSED** [review]
Looks good, but I seem to have overlooked a simple style adjustment in my previous preview. Sorry about that.
I will already commit the suggested fix, but please squash (and rebase if needed) once again. You can name the final commit as something like the title of the PR.

**Comment by davidBar-On** [comment]
Done - squashed and rebased.  Now with a meaningful squashed commit name.

**Comment by davidBar-On** [comment]
Done - squashed and rebased.  Now with a meaningful squashed commit name.

**Comment by davidBar-On** [comment]
Done - squashed and rebased.  Now with a meaningful squashed commit name.

File: libsolutil/Address.cpp
Comment (r0qs) [inline libsolutil/Address.cpp:36]: Although I'm ok with the original style, I'm just wondering if all messages should be formatted as the one at line 42. I guess this was the idea initially suggested by @matheusaaguiar here https://github.com/ethereum/solidity/pull/14283#discussion_r1213526671, right?

```suggestion			
		message =
				"Library address \"" + 
				_addrString +
				"\" is not prefixed with \"0x\".\n"
				"Note that the address must be prefixed with \"0x\".";
```

File: libsolutil/Address.cpp
Comment (r0qs) [inline libsolutil/Address.cpp:42]: Also, why not already return the error when it occurs instead of checking for `message.empty()` at the end? It would make the code more linear to read avoiding the nested `if`. For example:
```c++
Result<h160> validateAddress(string _addrString)
{
	string message;
	h160 address;

	if (_addrString.substr(0, 2) != "0x")
		return Result<h160>::err(
			"Library address \"" +
			_addrString +
			"\" is not prefixed with \"0x\".\n"
			"Note that the address must be prefixed with \"0x\".");

	string addrString = _addrString.substr(2);

	if (addrString.length() != 40)
		return Result<h160>::err(
			"Invalid library address length " +
			to_string(addrString.length()) +
			" instead of 40 characters.");

	if (!passesAddressChecksum(addrString, false))
		return Result<h160>::err(
			"Invalid checksum for library address \"" +
			_addrString +
			"\".\n"
			"The correct checksum is \"" +
			getChecksummedAddress(addrString) +
			"\".");

	try
	{
		address = util::h160(_addrString);
	}
	catch (util::BadHexCharacter const&)
	{
		return Result<h160>::err(
			"Invalid library address (\"" +
			_addrString +
			"\") supplied.");
	}

	return address;
}
```

**Comment by r0qs** [comment]
Although it may be redundant, this https://github.com/ethereum/solidity/issues/10299#issuecomment-874826732 in the issue pointed out that:

> CLI checks if the address is empty but Standard JSON does not. This check is redundant because we still check address length later but if it's there, it should be in both places.

So we should add a test for empty address in the Standard JSON as well.

File: solc/CommandLineParser.cpp
Comment (r0qs) [inline solc/CommandLineParser.cpp:444]: Oh actually the test for empty address was removed from the command-line so maybe it is fine indeed to also not have it in the Standard JSON, right @cameel? Since the address length is already  checked.

**Comment by r0qs** [comment]
By the way, thanks a lot for the contribution @davidBar-On, but to let you know, as the change in this PR is a breaking change we may not merge it right now into the develop branch. Please see https://github.com/ethereum/solidity/issues/10299#issuecomment-874837684 for context.

We still need to update our `breaking` branch with the latest develop and decide on what will be included in the next breaking release of the compiler, and this may take some time.

File: libsolutil/Address.cpp
Comment (davidBar-On) [reply libsolutil/Address.cpp:42]: I am used to try to minimize the number of function returns, unless it complicates the function readability too much.  This is to make sure that if there is a common code that is required before the return (releasing resources, debug message, etc.), it will not be repeated and there will be less chance that it will forgotten before one of the returns.

I did notice that that in this project the common practice is to return an error immediately, so I have no problem making the change.

File: libsolutil/Address.cpp
Comment (matheusaaguiar) [reply libsolutil/Address.cpp:36]: Well, I didn't mind with that one (and didn't point it specifically) because it's only the last part of the string literal and it is not really split or combined with an operator. For me it's like the second line here is just the first one "overlapping", if that makes sense.
For me it's fine either way.

File: solc/CommandLineParser.cpp
Comment (matheusaaguiar) [reply solc/CommandLineParser.cpp:444]: I missed that in my review. Although this is checked later, I think it is simple enough that the redundancy is harmless and worth the extra safety.

Comment (davidBar-On) [reply solc/CommandLineParser.cpp:444]: @ matheusaaguiar, just to be on the safe side - I assume that I should make the changes according to this comments.  Is this correct?  I am asking to make sure that this is not an internal discussion.

Comment (davidBar-On) [reply solc/CommandLineParser.cpp:444]: @matheusaaguiar, just to be on the safe side - I assume that I should make the changes according to this comments. Is this correct? I am asking to make sure that this is not an internal discussion.

Comment (r0qs) [reply solc/CommandLineParser.cpp:444]: @matheusaaguiar since @davidBar-On is already checked the address length here: https://github.com/ethereum/solidity/pull/14283/files#diff-38ee70850074a54efd416d53701471fa79742fd73f95e390e020beb01fdac93bR41 and we are now using the same `validateAddress` function that he created for both command-line and standard JSON, I think it is not necessary to check for empty address. But yeah, I just pointed it out so we are all aware that the check for an empty address was removed.

@davidBar-On I don't think you need to do anything in this regard. But let's see what @cameel has to say, since he was the one that mentioned such a check.

**Comment by ekpyron** [comment]
As also stated in https://github.com/ethereum/solidity/pull/14520#issuecomment-1698962069, unfortunately, we cannot simply *hope* that this won't affect any existing json input - this is a breaking change, so we need to treat it accordingly. Our ``breaking`` branch unfortunately is currently unmaintained and will need some work to make it feasible again to be targetted, so I'll close this PR for the time being. We can revisit when we revive the breaking branch.

**Pull request closed without merging**
